File: park-yeong-ki/6주차/PGS_양궁대회.java
import java.util.*;

class Solution {
    static int[] ryan, apeach, ans;
    static int max;
    static boolean flag;
    public int[] solution(int n, int[] info) {
        ryan = new int[11];
        apeach = info;
        max = 0;
        ans = new int[]{-1};
        
        dfs(0, n);
        
        return ans;
    }
            
    static void dfs(int idx, int cnt){
        if(idx == 10){
            ryan[idx] = cnt;
            
            int a = 0;
            int r = 0;
            boolean flag = false;
            for(int i=0; i<=10; i++){
                if(max > 0){
                    if(ans[i] < ryan[i]) flag = true;
                    else if(ans[i] > ryan[i]) flag = false;
                }
                
                if(apeach[i] == 0 && ryan[i] == 0) continue;
                if(apeach[i] < ryan[i]) r += 10 - i;
                else a += 10 - i;
            }
            
            if(max < r - a){
                max = r - a;
                ans = ryan.clone();
            }else if(max == r - a && flag){
                ans = ryan.clone();
            }
            
            return;
        }
        
        for(int i=0; i<=cnt; i++){
            ryan[idx] = i;
            dfs(idx+1, cnt-i);
            ryan[idx] = 0;
        }
    }
}

File: daehwa_park/10주차/PGS_양궁대회.py
nmax = 0
ans = [-1]

def solution(n, info):
    global ans
    
    ret = [0 for _ in range(11)]
    dfs(info, n, ret, 0)
    answer = ans
    
    return answer

def dfs(info, n, ret, cnt):
    global nmax, ans
    if n == 0 or cnt == 11:
        #계산
        ap = 0
        li = 0
        for i in range(11):
            if ret[i] == 0 and info[i] == 0: # 둘다 못맞힘
                continue
            elif ret[i] > info[i]: # 라이언 승
                li += 10 - i
            else: # 어피치 승
                ap += 10 - i
        if nmax < li - ap:
            nmax = li - ap
            ans = ret[:]
        elif nmax and nmax == li - ap:
            for i in range(10, -1, -1):
                if ret[i] == ans[i]:
                    continue
                elif ret[i] > ans[i]:
                    ans = ret[:]
                    break
                else:
                    break
        return
    
    if cnt == 10:
        ret[cnt] = n
        dfs(info, 0, ret, cnt + 1)
    elif n >= info[cnt] + 1:
        ret[cnt] = info[cnt] + 1
        dfs(info, n - (info[cnt] + 1), ret, cnt + 1) # 이겨
    ret[cnt] = 0
    dfs(info, n, ret, cnt + 1) # 져